### PR TITLE
Collect history of responses if redirects occur.

### PR DESCRIPTION
**Pull request by knitori**
Hi,

Accessing the history, or the information that a redirect occured at all (except for comparing original and final URL), is something I couldn't find and was missing from the requests library.

I'm completely new to this project, so I have no idea, if what I did, is the correct way to do it or if I missed something.

Feedback appreciated.

### CODE REVIEW

**Comment by asvetlov** [comment]
I support the PR but tests and documentation update should be added.
Would you manage the request up to be ready for merging?

**Comment by nitori** [comment]
I'll ... try? (I have never really done these things)

**Comment by asvetlov** [comment]
Please do. I'll help you.
Tests first.
Add new test to https://github.com/KeepSafe/aiohttp/blob/master/tests/test_client_functional.py
Create a test which has two server handlers: one for redirection and second for returning 200 OK response.
Check for `history` attribute of given client response.

**Comment by nitori** [comment]
Oh, I didn't see your comment until after I pushed the update. :-)

**Comment by asvetlov** [comment]
Github doesn't send email notification on pushing.
So please add a comment if you need feedback.

File: docs/client_reference.rst [hunk 1]
```diff
@@ -954,6 +954,11 @@ Response object
 
       HTTP headers of response, :class:`CIMultiDictProxy`.
 
+   .. attribute:: history
+
+      List of :class:`ClientResponse` objects of preceding requests, if there
```
Comment (asvetlov) [inline docs/client_reference.rst]: `:class:`list`` is even better

Comment (nitori) [inline docs/client_reference.rst]: Instead of "List" in "List of :class:`ClientResponse` objects …"?

Or where exactly do you mean?

Comment (asvetlov) [inline docs/client_reference.rst]: I mean 

```
:class:`list` of :class:`ClientResponse` objects of preceding requests ...
```

**Comment by nitori** [comment]
Okay, I added a test.

**Edit:** maybe add try/finally block around the asserts to release resp/resp_redirect?

**Comment by asvetlov** [comment]
Perfect!

> maybe add try/finally block around the asserts to release resp/resp_redirect?

Please do.

**Comment by nitori** [comment]
Done :-)

**Comment by asvetlov** [comment]
Let me sleep on it and merge tomorrow.

**Comment by popravich** [comment]
Hi,
The idea is good but there is an issue -- you simply collect responses into `history` list and then provide it with last non-redirecting response, howerer those responses (in the `history`) can not be accessed in full manner -- if someone try something like he will end with exception:

``` python
yield from resp.history[0].text()
```

**Comment by asvetlov** [comment]
Yes, all responses except the latest are released.
What do you suggest?
Requests has the same attribute IIRC.

**Comment by popravich** [comment]
I'm not quite sure what to do...
`resp.read() / json() / text()` coroutines use `content` attribute which is instance of some stream parser bound to `connection.reader` directly. And it all goes deeper into Connector/Connection stuff...
Need to look into this...

**Comment by asvetlov** [comment]
Good catch.
We may do better cleanup in `.close()/.release()`

Converted `.history` into readonly property by eac7028 BTW

**Comment by nitori** [comment]
If anything, I believe discarding the body and getting an exception when trying to access it should be the default behavior.

I assume that the common case is, that you're not even interested in the body of preceding requests (if anything, you're probably more interested in the headers or the URLs - as in my case), and in most cases, it doesn't contain any valuable information anyway.

So if you _happen_ to be interested in the body of those requests, there should either be an option to completely load them or to somehow keep the connection opened, or to use `allow_redirects=False` and go through the steps yourself.

**Comment by asvetlov** [comment]
> If anything, I believe discarding the body and getting an exception when trying to access it should be the default behavior.

Would you propose a patch?

**Comment by nitori** [comment]
Hm, I'd have to take a closer look at how things work, but I would like to give it a try.

As far as I can tell, awaiting .text() doesn't actually raise _any_ exception but just returns an empty string on released or closed responses. Changing that would probably cause some applications, that rely on that, to break, and I don't know if it's worth to add an exemption just for redirects.

I assumed it already raises some kind of exception - so I'm not sure after all, if it's a good idea.

**Comment by asvetlov** [comment]
I don't care about `.text()` right now but `.content` should be protected from further usage.

**Comment by nitori** [comment]
We might as well just close the preceding responses instead of [releasing](https://github.com/KeepSafe/aiohttp/blob/eac702825f257414e02a6796dab2d7b64b940f6b/aiohttp/client.py#L221) them. If we do that, trying to read from `.content` will cause an `aiohttp.errors.ServerDisconnectedError`.

**Comment by asvetlov** [comment]
No, response releasing is crucial for HTTP keep-alive support.
`.close()` call breaks connection and closes socket but `.release()` allows to reuse the socket for next request.

Replacing `.content` with something like `EmptyStream` (https://github.com/KeepSafe/aiohttp/blob/master/aiohttp/streams.py#L332) but raises `RuntimeError` on every read call may prevent reading from history responses. 

Concurrent reading from different contents (bound with the same underlying transport) make a mess as @popravich mentioned.

Honestly using just `EmptyStream` is an option also: redirection responses stored in history should have no HTTP BODY anyway.

Did you get my point?

**Pull request merged**